src/code/z_message.c

```diff
@@ -64,46 +322,246 @@
+        }
+    }
+}
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80151DA4.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80152434.s")
+void func_80152434(GlobalContext* globalCtx, u16 arg2) {
+    globalCtx->msgCtx.unk12046 = 0;
+    func_80151DA4(globalCtx, arg2);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80152464.s")
+void func_80152464(GlobalContext* globalCtx, u16 arg1) {
+    globalCtx->msgCtx.unk12046 = 1;
+    func_80151DA4(globalCtx, arg1);
+}
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/Message_GetState.s")
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_8015268C.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80152C64.s")
+void func_80152C64(View* view) {
+    Viewport viewport;
+    // clang-format off
+    viewport.bottomY = 0xF0;viewport.rightX = 0x140;
+    // clang-format on
+    viewport.topY = 0;
+    viewport.leftX = 0;
+
+    View_SetViewport(view, (Viewport*)&viewport);
+    func_8013FBC8(view);
+}
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80152CAC.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80152EC0.s")
+// Spawn song effect?
+void func_80152EC0(GlobalContext* globalCtx) {
+    MessageContext* msgCtx = &globalCtx->msgCtx;
+    Actor* actor = (Actor*)GET_PLAYER(globalCtx);
+
+    if (1) {}
+    if ((msgCtx->songPlayed < 0x17) && (msgCtx->songPlayed != 0xE) &&
+        ((msgCtx->ocarinaAction < 0x43) || (msgCtx->ocarinaAction >= 0x47))) {
+        msgCtx->unk120B0 = 1;
+        if (msgCtx->songPlayed != 0x16) {
+            Actor_Spawn(&globalCtx->actorCtx, globalCtx, D_801D02D8[msgCtx->songPlayed], actor->world.pos.x,
+                        actor->world.pos.y, actor->world.pos.z, 0, 0, 0, D_801D02F8[msgCtx->songPlayed]);
+            return;
+        }
+        Actor_Spawn(&globalCtx->actorCtx, globalCtx, 0xF6, actor->world.pos.x, actor->world.pos.y, actor->world.pos.z,
+                    0, 0, 0, 0);
+    }
+}
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80152FB8.s")
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80153750.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80153E7C.s")
+void func_80153E7C(GlobalContext* globalCtx, void* arg1) {
+    if ((gSaveContext.language == 0) && (globalCtx->msgCtx.unk12090 == 0)) {
+        func_8014ADBC(globalCtx, arg1);
+        return;
+    }
+    if (globalCtx->msgCtx.unk12090 != 0) {
+        func_8015E7EC(globalCtx, arg1);
+        return;
+    }
+    func_8015966C(globalCtx, arg1, 0);
+}
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80153EF0.s")
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_801541D4.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80156758.s")
+void func_80156758(GlobalContext* globalCtx) {
+    Gfx* nextDisplayList;
+    Gfx* _polyOpa;
+    GraphicsContext* gfxCtx;
+
+    gfxCtx = globalCtx->state.gfxCtx;
+    OPEN_DISPS(gfxCtx);
```

Comment:
We usually try to avoid prefixing variables with an underscore
```suggestion
    Gfx* nextDisplayList;
    Gfx* polyOpa;
    GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;

    OPEN_DISPS(gfxCtx);
```

Reply:
done in acd12b2b